treasury.go

```diff
@@ -0,0 +1,194 @@
+package dcrlibwallet
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"decred.org/dcrwallet/v2/errors"
+
+	"github.com/decred/dcrd/blockchain/stake/v4"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+)
+
+// SetTreasuryPolicy saves the voting policy for treasury spends by a particular
+// PI key.
+// If a ticket hash is provided, the voting policy is also updated with the VSP
+// controlling the ticket. If a ticket hash isn't provided, the vote choice is
+// saved to the local wallet database and the VSPs controlling all unspent,
+// unexpired tickets are updated to use the specified vote policy.
+func (wallet *Wallet) SetTreasuryPolicy(PiKey, newVotingPolicy, tixHash string, passphrase []byte) error {
+	var ticketHash *chainhash.Hash
+	if tixHash != "" {
+		tixHash, err := chainhash.NewHashFromStr(tixHash)
+		if err != nil {
+			return fmt.Errorf("inavlid ticket hash: %w", err)
+		}
+		ticketHash = tixHash
+	}
+
+	// The wallet will need to be unlocked to sign the API
+	// request(s) for setting this voting policy with the VSP.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
```

Comment:
Perform this unlocking after the pikey and newVotingPolicy checks. No need to unlock the wallet and immediately relock it if either the pikey or newVotingPolicy is invalid.